File: Client/src/api/services/workoutService.ts
import { requests } from '@/api/api';
import { Workout } from '@/types/Workout';

const workoutService = {
  getWorkoutsForTrainee: async (traineeId: number) => {
    const workouts = await requests.get<Workout[]>(
      `/workout/trainee/${traineeId}`,
    );
    return workouts;
  },
  getWorkoutForTrainee: async (traineeId: number, workoutId: number) => {
    const workout = await requests.get<Workout>(
      `/workout/trainee/${traineeId}/${workoutId}`,
    );
    return workout;
  },
  createWorkout: async (workout: Workout) => {
    const newWorkout = await requests.post<Workout>(`/workout`, workout);
    return newWorkout;
  },
};

export default workoutService;

File: backend/Easygym.Application/Services/WorkoutService.cs
using Easygym.Domain.Constants;
using Easygym.Domain.Entities;
using Easygym.Domain.Interfaces;
using Easygym.Domain.Exceptions;
using Easygym.Domain.Models.Requests;
namespace Easygym.Application.Services
{
    public class WorkoutService
    {
        private readonly IWorkoutRepository _workoutRepository;
        private readonly CurrentUserService _currentUserService;

        public WorkoutService(IWorkoutRepository workoutRepository, CurrentUserService currentUserService)
        {
            _workoutRepository = workoutRepository;
            _currentUserService = currentUserService;
        }
        public async Task<List<Workout>> GetWorkoutsForTraineeAsync(int traineeId)
        {
            await CanAccessWorkout(traineeId);

            var workouts = await _workoutRepository.GetWorkoutsForTraineeAsync(traineeId);
            return workouts;
        }

        public async Task<Workout> GetWorkoutForTraineeAsync(int workoutId, int traineeId)
        {
            await CanAccessWorkout(traineeId);

            var workout = await _workoutRepository.GetWorkoutForTraineeAsync(workoutId, traineeId);
            return workout;
        }

        public async Task<Workout> CreateWorkoutAsync(Workout workout)
        {
            await CanAccessWorkout(workout.TraineeId);

            if (workout.Sets.Count == 0)
            {
                throw new ValidationException("Workout must have at least one set");
            }

            await _workoutRepository.AddWorkoutAsync(workout);
            return workout;
        }


        public async Task<Workout> UpdateWorkoutAsync(int traineeId, int workoutId, UpdateWorkoutRequest workout)
        {
            await CanAccessWorkout(traineeId);
            return await _workoutRepository.UpdateWorkoutAsync(workoutId, workout);
        }


        public async Task DeleteWorkoutAsync(int traineeId, int workoutId)
        {
            await CanAccessWorkout(traineeId);
            await _workoutRepository.DeleteWorkoutAsync(workoutId);
        }

        public async Task CanAccessWorkout(int traineeId)
        {
            var currentUser = await _currentUserService.GetCurrentUserAsync();

            // Only allow trainers and admins to see other users' workouts,
            // while clients can only see their own workouts
            if (currentUser.Role == Role.Client && currentUser.Id != traineeId)
            {
                throw new ForbiddenAccessException();
            }
        }
    }
}